servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -4770,6 +4792,14 @@ public Assertion putAssertion(ResourceContext ctx, String domainName, String pol
 
         validatePolicyAssertion(assertion, caller);
 
+        // we are not going to allow any user to add/update
+        // an assertion that the associated role does not exist
+
+        RoleList roleList = getRoleList(ctx, domainName, null, null);
```

Comment:
1) Take a look at implementing this functionality within validatePolicyAssertions/validatePolicyAssertion calls.

2) the logic with if (roleList != null && roleList.getNames() != null) is actually incorrect since if the domain for some reason has no role then we're allowing it to add any assertion without any checks. If the domain has no roles then we must always reject all assertions since there is something wrong with the domain.